StatBot/EchoBot.cs

```diff
@@ -23,49 +36,206 @@ public class EchoBot : IBot
+                            .Where(types => types.IsClass && !types.IsAbstract && types.GetInterface("ICard") != null);
+
+
+                        foreach (var botButton in botButtons)
+                        {
+                            _cards.Add((ICard)Activator.CreateInstance(botButton, StatisticId, null,null));
+                        }
 
-                    var help = new Help();
+                        var str_ = ButtonType;
 
-                    var tool = _tools.FirstOrDefault(x => x.CommandsName.Any(y => y.Equals(command)));
-                    if (tool != null)
+                        var card = _cards.FirstOrDefault(x => x.ButtonsName.Any(y => y.Equals(str_)));
+                        if (card != null)
+                        {
+                            await context.SendActivity(card.OnClick(context.Activity));
+                        }
+                    }
+                    else
                     {
-                        context.Activity.Text = indexOfSpace >= 0 ? context.Activity.Text.Substring(indexOfSpace, str.Length - indexOfSpace) : String.Empty;
-                        if (user == null || (!user.IsAdmin && ((ITool)tool).IsAdmin))
+                        var user = DataModel.RememberUser(context.Activity);
+                        List<ITool> _tools = new List<ITool>();
+
+
+                        var baseInterfaceType = typeof(ITool);
+                        var botCommands = Assembly.GetAssembly(baseInterfaceType)
+                            .GetTypes()
+                            .Where(types => types.IsClass && !types.IsAbstract && types.GetInterface("ITool") != null);
+                        foreach (var botCommand in botCommands)
+                        {
+                            _tools.Add((ITool)Activator.CreateInstance(botCommand));
+                        }
+
+                        var str = context.Activity.Text.Trim();
+                        var indexOfSpace = str.IndexOf(" ", StringComparison.Ordinal);
+                        var command = indexOfSpace != -1 ? str.Substring(0, indexOfSpace).ToLower() : str.ToLower();
+                        if (command[0] != '/')
                         {
-                            await context.SendActivity(help.Run(context.Activity));
-                            return;
+                            command = "/" + command;
                         }
-                        await context.SendActivity(tool.Run(context.Activity));
+
+                        var help = new Help();
+
+                        var tool = _tools.FirstOrDefault(x => x.CommandsName.Any(y => y.Equals(command)));
+                        if (tool != null)
+                        {
+                            context.Activity.Text = indexOfSpace >= 0 ? context.Activity.Text.Substring(indexOfSpace + 1, str.Length - indexOfSpace - 1) : String.Empty;
+                            if (user == null || (!user.IsAdmin && ((ITool)tool).IsAdmin))
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this unnecessary cast to 'ITool'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://172.21.13.172:9091/coding_rules#rule_key=csharpsquid%3AS1905)